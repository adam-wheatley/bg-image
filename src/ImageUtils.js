import { convertProps } from './HelperUtils'


const imageCache = {}
/**
 * Cache if we've seen an image before so we don't both with
 * lazy-loading & fading in on subsequent mounts.
 *
 * @param props
 * @return {*|boolean}
 */
export const inImageCache = props => {
  const convertedProps = convertProps(props)
  // Find src
  const src = convertedProps.fluid
      ? convertedProps.fluid.src
      : convertedProps.fixed.src

  return imageCache[src] || false
}


/**
 * Adds an Image to imageCache.
 * @param props
 */
export const activateCacheForImage = props => {
  const convertedProps = convertProps(props)
  // Find src
  const src = convertedProps.fluid
      ? convertedProps.fluid.src
      : convertedProps.fixed.src

  imageCache[src] = true
}


/**
 * Creates a picture element for the browser to decide which image to load.
 *
 * @param props
 * @param onLoad
 * @return {HTMLImageElement|null}
 */
export const createPictureRef = (props, onLoad = () => {}) => {
  const convertedProps = convertProps(props)
  if (typeof window !== `undefined` &&
      (typeof convertedProps.fluid !== `undefined` ||
       typeof convertedProps.fixed !== `undefined`)) {
    const imageData = props.fluid
        ? props.fluid
        : props.fixed

    const img = new Image()
    const pic = document.createElement('picture')
    if (imageData.srcSetWebp) {
      const sourcesWebP = document.createElement('source')
      sourcesWebP.type = `image/webp`
      sourcesWebP.srcset = imageData.srcSetWebp
      sourcesWebP.sizes = imageData.sizes
      pic.appendChild(sourcesWebP)
    }
    pic.appendChild(img)

    img.onload = () => onLoad()
    if (!img.complete && typeof convertedProps.onLoad === `function`) {
      img.addEventListener('load', convertedProps.onLoad)
    }
    if (typeof convertedProps.onError === `function`) {
      img.addEventListener('error', convertedProps.onError)
    }
    img.srcset = imageData.srcSet ? imageData.srcSet : ``
    img.src = imageData.src ? imageData.srcSet : ``

    return img
  }
  return null
}


/**
 * Create basic image for a noscript event.
 *
 * @param props
 * @return {string}
 */
export const noscriptImg = props => {
  // Check if prop exists before adding each attribute to the string output below to prevent
  // HTML validation issues caused by empty values like width="" and height=""
  const src = props.src ? `src="${props.src}" ` : `src="" ` // required attribute
  const sizes = props.sizes ? `sizes="${props.sizes}" ` : ``
  const srcSetWebp = props.srcSetWebp
      ? `<source type='image/webp' srcset="${props.srcSetWebp}" ${sizes}/>`
      : ``
  const srcSet = props.srcSet ? `srcset="${props.srcSet}" ` : ``
  const title = props.title ? `title="${props.title}" ` : ``
  const alt = props.alt ? `alt="${props.alt}" ` : `alt="" ` // required attribute
  const width = props.width ? `width="${props.width}" ` : ``
  const height = props.height ? `height="${props.height}" ` : ``
  const opacity = props.opacity ? props.opacity : `1`
  const transitionDelay = props.transitionDelay ? props.transitionDelay : `0.5s`
  return `<picture>${srcSetWebp}<img ${width}${height}${sizes}${srcSet}${src}${alt}${title}style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:${transitionDelay};opacity:${opacity};width:100%;height:100%;object-fit:cover;object-position:center"/></picture>`
}