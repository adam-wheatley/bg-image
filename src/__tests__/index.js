import '@babel/polyfill'
import { render, cleanup, fireEvent } from 'react-testing-library'
import React from 'react'
import BackgroundImage from '../'
import { createPictureRef } from '../ImageUtils'


global.console.debug = jest.fn()

afterAll(cleanup)


export const fixedShapeMock = {
  width: 100,
  height: 100,
  src: `test_fixed_image.jpg`,
  srcSet: `some srcSet`,
  srcSetWebp: `some srcSetWebp`,
  base64: `string_of_base64`,
}

export const fluidShapeMock = {
  aspectRatio: 1.5,
  src: `test_fluid_image.jpg`,
  srcSet: `some srcSet`,
  srcSetWebp: `some srcSetWebp`,
  sizes: `(max-width: 600px) 100vw, 600px`,
  base64: `string_of_base64`,
}

/**
 * Sets up a (react-testing-library) rendered container.
 *
 * @param fluid
 * @param addClass
 * @param additionalClass
 * @param fixedClass
 * @param onLoad
 * @param onError
 * @param critical
 * @param onStartLoad
 * @return {RenderResult.container}
 */
const setup = (fluid = false,
               addClass = false,
               additionalClass = ``,
               fixedClass = true,
               onLoad = () => {},
               onError = () => {},
               critical = false,
               onStartLoad = null) => {

  if (addClass) {
    // Create the style class.
    const styleElement = document.createElement("style")
    styleElement.textContent = `
      .fixedImage {
        backgroundRepeat: 'repeat-y';
      }
    `
    document.body.appendChild(styleElement)
  }
  const classNames = fixedClass ?
      `fixedImage${additionalClass}` : additionalClass.trim()
  const { container } = render(
    <BackgroundImage
      backgroundColor
      className={classNames}
      style={{ display: `inline`, opacity: .99 }}
      title={`Title for the image`}
      alt={`Alt text for the image`}
      {...fluid && { fluid: fluidShapeMock }}
      {...!fluid && { fixed: fixedShapeMock }}
      onLoad={onLoad}
      onError={onError}
      placeholderStyle={{ color: `red` }}
      placeholderClassName={`placeholder`}
      classId="test"
      critical={critical}
      onStartLoad={onStartLoad}
    ><h1>test</h1></BackgroundImage>
  )

  return container
}

describe(`<BackgroundImage />`, () => {
  beforeEach(() => {
    global.IntersectionObserver = jest.fn(() => ({
      observe: jest.fn(),
    }))
  })

  it(`should render fixed size images`, () => {
    const component = setup()
    expect(component).toMatchSnapshot()
  })

  it(`should call critical fixed images`, () => {
    const component = setup(false, true, ``, true, null, null, true)
    expect(component).toMatchSnapshot()
  })

  it(`should render fluid images`, () => {
    const component = setup(true)
    expect(component).toMatchSnapshot()
  })

  it(`should call critical fluid images`, () => {
    const component = setup(true, true, ``, true, null, null, true)
    expect(component).toMatchSnapshot()
  })

  it(`should work without external class`, () => {
    // Mock Math.random beforehand, lest another random classname is created.
    Math.random = jest.fn(() => 0.424303425546642)
    const { container } = render(
        <BackgroundImage
            fluid={ fluidShapeMock }
        ><h1>testempty</h1></BackgroundImage>
    )
    expect(container).toMatchSnapshot()
  })

  it(`should change style.display from 'inherit' to 'inline-block'`, () => {
    // Mock Math.random beforehand, lest another random classname is created.
    Math.random = jest.fn(() => 0.424303425546642)
    const { container } = render(
        <BackgroundImage
            fixed={ fixedShapeMock }
            style={{ display: `inherit` }}
        ><h1>testempty</h1></BackgroundImage>
    )
    expect(container).toMatchSnapshot()
  })

  it(`should work with single external class`, () => {
    const component = setup(true)
    expect(component).toMatchSnapshot()
  })

  it(`should work with another external class`, () => {
    const component = setup(true, true, ` test`)
    expect(component).toMatchSnapshot()
  })

  it(`should create style tag with pseudo-elements`, () => {
    const styleTag = setup(true, true, ` test`).querySelector(`style`)
    expect(styleTag).toBeInTheDocument()
    expect(styleTag).toMatchSnapshot()
  })

  it(`should create style tag at fixed with pseudo-elements`, () => {
    const styleTag = setup(false, true, ` test`).querySelector(`style`)
    expect(styleTag).toBeInTheDocument()
    expect(styleTag).toMatchSnapshot()
  })

  it(`should have class with pseudo element in style tag`, () => {
    const styleTag = setup(true, true, ` test`).querySelector(`style`)
    expect(styleTag)
        .toHaveTextContent(`.gatsby-background-image-test:before`)
    expect(styleTag)
        .toHaveTextContent(`background-repeat: 'repeat-y';`)
  })

  it(`should call onLoad and onError image events`, () => {
    const onLoadMock = jest.fn()
    const onErrorMock = jest.fn()
    const image = createPictureRef({
      fluid: {
        src: ``,
      },
      onLoad: onLoadMock,
      onError: onErrorMock
    }, onLoadMock)
    fireEvent.load(image)
    fireEvent.error(image)

    expect(onLoadMock).toHaveBeenCalledTimes(2)
    expect(onErrorMock).toHaveBeenCalledTimes(1)
  })
})


describe(`<BackgroundImage /> without IO`, () => {
  beforeEach(() => {
    delete global.IntersectionObserver
  })

  it(`should call onLoadFunction without IO`, () => {
    const onLoadFunctionMock = jest.fn()
    const component = setup(true, true, ``, true, null, null, true, onLoadFunctionMock)
    expect(component).toMatchSnapshot()
    expect(onLoadFunctionMock).toHaveBeenCalled()
  })
})
