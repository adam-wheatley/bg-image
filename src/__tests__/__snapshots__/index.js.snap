// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackgroundImage /> should call critical fixed images 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; display: inline; width: 100px; height: 100px; opacity: 0.99;"
  >
    <style>
      
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: 'repeat-y';
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(test_fixed_image.jpg);
            background-repeat: 'repeat-y';
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" /&gt;&lt;img width="100" height="100" srcset="some srcSet" src="test_fixed_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should call critical fluid images 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; display: inline;"
  >
    <style>
      
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: 'repeat-y';
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(test_fluid_image.jpg);
            background-repeat: 'repeat-y';
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img sizes="(max-width: 600px) 100vw, 600px" srcset="some srcSet" src="test_fluid_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should change style.display from 'inherit' to 'inline-block' 1`] = `
<div>
  <div
    class=" gatsby-background-image-fwatluf gatsby-image-wrapper"
    style="position: relative; overflow: hidden; width: 100px; height: 100px; opacity: 0.99;"
  >
    <style>
      
          .gatsby-background-image-fwatluf:before,
          .gatsby-background-image-fwatluf:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-fwatluf:before {
            z-index: -100;
            
            background-repeat: no-repeat;
            
            opacity: 0; 
          }
          .gatsby-background-image-fwatluf:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            background-repeat: no-repeat;
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" /&gt;&lt;img width="100" height="100" srcset="some srcSet" src="test_fixed_image.jpg" alt="" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      testempty
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should create style tag at fixed with pseudo-elements 1`] = `
<style>
  
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: 'repeat-y';
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(string_of_base64);
            background-repeat: 'repeat-y';
          }
        
</style>
`;

exports[`<BackgroundImage /> should create style tag with pseudo-elements 1`] = `
<style>
  
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: 'repeat-y';
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(string_of_base64);
            background-repeat: 'repeat-y';
          }
        
</style>
`;

exports[`<BackgroundImage /> should render fixed size images 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; display: inline; width: 100px; height: 100px; opacity: 0.99;"
  >
    <style>
      
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: no-repeat;
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(string_of_base64);
            background-repeat: no-repeat;
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" /&gt;&lt;img width="100" height="100" srcset="some srcSet" src="test_fixed_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should render fluid images 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; display: inline;"
  >
    <style>
      
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: 'repeat-y';
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(string_of_base64);
            background-repeat: 'repeat-y';
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img sizes="(max-width: 600px) 100vw, 600px" srcset="some srcSet" src="test_fluid_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work with another external class 1`] = `
<div>
  <div
    class="fixedImage test gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; display: inline;"
  >
    <style>
      
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: 'repeat-y';
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(string_of_base64);
            background-repeat: 'repeat-y';
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img sizes="(max-width: 600px) 100vw, 600px" srcset="some srcSet" src="test_fluid_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work with single external class 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; display: inline;"
  >
    <style>
      
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: 'repeat-y';
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(string_of_base64);
            background-repeat: 'repeat-y';
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img sizes="(max-width: 600px) 100vw, 600px" srcset="some srcSet" src="test_fluid_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> should work without external class 1`] = `
<div>
  <div
    class=" gatsby-background-image-fwatluf gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99;"
  >
    <style>
      
          .gatsby-background-image-fwatluf:before,
          .gatsby-background-image-fwatluf:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-fwatluf:before {
            z-index: -100;
            
            background-repeat: no-repeat;
            
            opacity: 0; 
          }
          .gatsby-background-image-fwatluf:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            background-repeat: no-repeat;
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img sizes="(max-width: 600px) 100vw, 600px" srcset="some srcSet" src="test_fluid_image.jpg" alt="" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      testempty
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> without IO should call onLoadFunction without IO 1`] = `
<div>
  <div
    class="fixedImage gatsby-background-image-test gatsby-image-wrapper"
    style="position: relative; overflow: hidden; opacity: 0.99; display: inline;"
  >
    <style>
      
          .gatsby-background-image-test:before,
          .gatsby-background-image-test:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            -webkit-background-size: cover;
            -moz-background-size: cover;
            -o-background-size: cover;
            background-size: cover;
            -webkit-transition-delay: 0.25s;
            -moz-transition-delay: 0.25s;
            -o-transition-delay: 0.25s;
            transition-delay: 0.25s;
            -webkit-transition: opacity 0.5s;
            -moz-transition: opacity 0.5s;
            -o-transition: opacity 0.5s;
            transition: opacity 0.5s;
          }
          .gatsby-background-image-test:before {
            z-index: -100;
            
            background-repeat: 'repeat-y';
            background-color: lightgray;
            opacity: 0; 
          }
          .gatsby-background-image-test:after {
            z-index: -101;
            background-color: lightgray;
            background-image: url(test_fluid_image.jpg);
            background-repeat: 'repeat-y';
          }
        
    </style>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcset="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img sizes="(max-width: 600px) 100vw, 600px" srcset="some srcSet" src="test_fluid_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
    <h1>
      test
    </h1>
  </div>
</div>
`;
